### DISCUSSION TITLE
Determining/specifying network interface for socket

### DISCUSSION THREAD

**Discussion by victorallume** [micropython/micropython#12590] [Libraries & Drivers]
If I have two network interfaces active (e.g. wlan and ethernet), how does micropython determine which one to use for a socket? Is it possible to specify for either an outgoing or incoming socket connection to use a particular interface?

**Comment by jimmo** [comment]
If you're using LWIP (e.g. pyboard-d) then `socket.bind` sets the interface.

**Reply by victorallume** [reply]
Isn't `socket.bind` for opening a listening socket? Can I specify the interface for an outgoing socket?

**Reply by jimmo** [reply]
You can use it for both.

**Reply by victorallume** [reply]
Thanks. How does `socket.getaddrinfo()` determine which interface to use?

**Reply by jimmo** [reply]
Ahh, good question. On CPython, that would be the same as BSD sockets getaddrinfo, which would say "that's up to the system resolver".

In our case, LWIP is figuring it out based on the configured dns servers available. Here's what I've learnt from the LWIP source code + MicroPythn bindings:
 - getaddrinfo calls LWIP dns_gethostbyname, which iterates all known dns servers, and for each server will use the first available interface that can route that address.
 - LWIP has a global set of dns servers not bound to the interface. When an interface gets a DHCP response, it can add the dns server to that list.
 - MicroPython's nic.ifconfig() only ever considers the zeroth LWIP dns server (for both set and get).

So if both your interfaces can route the configured DNS server, then both could be used. But if only one can, then it will do the right thing. What does your interface configuration look like?

**Reply by victorallume** [reply]
I've got one Wifi interface in AP mode, so it's not able to talk to the internet, and another Ethernet interface, which will (probably) be talking to the internet.
